File: project/TempCode/LSH.java
import java.io.Serializable;

public abstract class LSH implements Serializable {

    protected static final long LARGE_PRIME =  433494437;
    private static final int DEFAULT_BANDS = 3;
    private static final int DEFAULT_BUCKETS = 10;

    private int bands = DEFAULT_BANDS;
    private int buckets = DEFAULT_BUCKETS;

    //Instantiates a LSH instance with s  bands and b buckets (per band)
    public LSH(final int stages, final int buckets) {
        this.bands = stages;
        this.buckets = buckets;
    }

    //Hash a signature.
    public final int[] hashSignature(final int[] signature) {

    	//signature is for each object
        int[] hash = new int[bands];
        
        // Number of rows per stage
        int rows = signature.length / bands;
        
        for (int i = 0; i < signature.length; i++) {
            int stage = Math.min(i / rows, bands - 1);
            hash[stage] = (int)
                    ((hash[stage] + (long) signature[i] * LARGE_PRIME)
                    % buckets);

        }

        return hash;
    }
}

File: project/TL_SWAR_analysis/LSH.cpp
//
//  LSH.cpp
//  LLVMDemangle
//
//  Created by chauncey on 2018/4/9.
//
#include "LSH.hpp"
#include <algorithm>
using namespace SWAR_TL;
using namespace std;

int * LSH::hashSignature(int * signature, int sig_size){
    //signature is for each object
    int hash[bands];
    
    // Number of rows per stage
    int rows =  sig_size / bands;
    
    for (int i = 0; i < sig_size; i++) {
        int stage = min(i / rows, bands - 1);
        hash[stage] = (int)((hash[stage] + (long) signature[i] * LARGE_PRIME) % buckets);
    }
    return hash;
};
